ext/nnstreamer/tensor_filter/tensor_filter_lua.cc

```diff
@@ -74,13 +97,64 @@ void _fini_filter_lua (void) __attribute__((destructor));
 }
 #endif /* __cplusplus */
 
+/**
+ * @brief Data for lua tensor
+ */
+typedef struct lua_tensor {
+  tensor_type type;
+  void *data;
+  size_t size;
+} lua_tensor;
+
 /** @brief For getting value in Lua */
 static int
 tensor_index (lua_State *L)
 {
-  uint8_t** parray = (uint8_t **) luaL_checkudata (L, 1, "tensor");
-  int index = luaL_checkint (L, 2);
-  lua_pushnumber (L, (*parray)[index - 1]);
+  lua_tensor *lt = *((lua_tensor **) luaL_checkudata(L, 1, "lua_tensor"));
+  int tidx = luaL_checkint (L, 2) - 1;
+
+  uint element_size = gst_tensor_get_element_size (lt->type);
+  if (tidx < 0 || tidx * element_size >= lt->size)
+    throw std::runtime_error ("Invalid index for tensor");
+
+  double value = 0.0;
+  switch (lt->type) {
```

Comment:
can you replace and test these using gst_tensor_data_raw_typecast() ?
```
if (!gst_tensor_data_raw_typecast (lt->data + (element_size * tidx), lt->type, &value, _NNS_FLOAT64)) {
  throw std::runtime_error ("Error occured during get tensor value");
}
```

Reply:
plz consider line 174 also.

Reply:
ditto. It reduces CC, overhead, and unreadability :)

Reply:
There are some concerns about the performance. With simple copy-paste lua model, I tested below and checked the total time of function `tensor_index` consumes:

```bash
$ gst-launch-1.0 videotestsrc num-buffers=100 ! videoconvert ! \
 videoscale ! video/x-raw,width=299,height=299,format=RGB ! \
tensor_converter ! tensor_filter framework=lua model=../passthrough.lua ! \ 
tensor_sink
```

With original code:
```bash
** Message: 16:41:00.976: tensor_index uses time: 1871ms
Got EOS from element "pipeline0".
Execution ended after 0:00:05.505801078
Setting pipeline to PAUSED ...
Setting pipeline to READY ...
Setting pipeline to NULL ...
Freeing pipeline ...
```

With `gst_tensor_data_raw_typecast`:
```bash
** Message: 16:45:05.997: tensor_index uses time: 2218ms
Got EOS from element "pipeline0".
Execution ended after 0:00:05.981997883
Setting pipeline to PAUSED ...
Setting pipeline to READY ...
Setting pipeline to NULL ...
Freeing pipeline ...
```

1871 ms vs 2218 ms seems quite a big difference to me.
Once optimize `gst_tensor_data_raw_typecast`, then it could be applied here.

Reply:
Then, plz consider data conversion in tensor_newindex() (double > unsigned vs double > signed > unsigned)

Reply:
Thanks! I updated the code. PTAL